egs/wsj/s5/steps/dict/learn_lexicon.sh

```diff
@@ -0,0 +1,402 @@
+    $dir/phone_decode/prons.txt $dir/phone_decoding_generated_prons_unfiltered.txt
+  cat $dir/phone_decoding_generated_prons_unfiltered.txt | grep -vP "<eps>|<UNK>|<unk>|\[.*\]" | \
+    sort | uniq > $dir/phone_decoding_generated_prons.txt || exit 1;
+fi
+
+if [ $stage -le 4 ]; then
+  echo "$0: Combine the reference lexicon and pronunciations from phone-decoding/G2P into one"
+  echo "  ... lexicon, and run lattice alignment using this lexicon on acoustic training data."
+  
+  mkdir -p $dir/dict_combined
+  cp $ref_dict/{extra_questions.txt,optional_silence.txt,nonsilence_phones.txt,silence_phones.txt} \
+    $dir/dict_combined/ 2>/dev/null
+  rm $dir/dict_combined/lexiconp.txt 2>/dev/null
+
+  cat $dir/phone_decoding_generated_prons.txt $dir/g2p_pron_candidates.txt \
+    $dir/ref_lexicon.txt | tr -s '\t' ' ' | \
+    awk 'NR==FNR{a[$1] = 1; next} ($1 in a)' $dir/train_counts.txt - | \
+    cat $dir/non_scored_entries - | \
+    sort | uniq > $dir/dict_combined/lexicon.txt
+  
+  utils/prepare_lang.sh --phone-symbol-table $ref_lang/phones.txt \
+    $dir/dict_combined $oov_symbol \
+    $dir/lang_combined_local $dir/lang_combined || exit 1;
+  
+  mdl=$src_mdl && $retrain_src_mdl && mdl=${src_mdl}_${affix}
+  steps/align_fmllr_lats.sh --cmd "$decode_cmd" --nj $nj \
+    $data $dir/lang_combined $mdl $dir/lats || exit 1;
+
+  # Get arc level information from the lattice.
+  $cmd JOB=1:$nj $dir/lats/log/get_arc_info.JOB.log \
+    lattice-align-words $dir/lang_combined/phones/word_boundary.int \
+    $dir/lats/final.mdl \
+    "ark:gunzip -c $dir/lats/lat.JOB.gz |" ark:- \| \
+    lattice-arc-post --acoustic-scale=0.1 $dir/lats/final.mdl ark:- - \| \
+    utils/int2sym.pl -f 5 $dir/lang_combined/words.txt \| \
+    utils/int2sym.pl -f 6- $dir/lang_combined/phones.txt '>' \
+    $dir/lats/arc_info_sym.JOB.txt || exit 1;
+  
+  # Get soft counts of all pronunciations from arc level information.
+  cat $dir/lats/arc_info_sym.*.txt | steps/dict/get_pron_stats.py - \
+    $dir/phone_decode/phone_map.txt $dir/lats/pron_stats.txt || exit 1;
+fi
+
+if [ $stage -le 5 ]; then
+  echo "$0: Prune the pronunciation candidates generated from G2P/phone-decoding, and re-do lattice-alignment."
+  mkdir -p $dir/dict_combined2
+  cp $ref_dict/{extra_questions.txt,optional_silence.txt,nonsilence_phones.txt,silence_phones.txt} \
+    $dir/dict_combined2/ 2>/dev/null
+  rm $dir/dict_combined2/lexiconp.txt 2>/dev/null
+
+  steps/dict/prune_pron_candidates.py $dir/lats/pron_stats.txt $dir/ref_lexicon.txt $dir/pruned_prons.txt
+ 
+  awk 'NR==FNR{a[$0] = 1; next} (!($0 in a))' $dir/pruned_prons.txt $dir/phone_decoding_generated_prons.txt \
+    > $dir/phone_decoding_generated_prons_pruned.txt
+
+  awk 'NR==FNR{a[$0] = 1; next} (!($0 in a))' $dir/pruned_prons.txt $dir/g2p_pron_candidates.txt \
+    > $dir/g2p_pron_candidates_pruned.txt \
+
+  cat $dir/phone_decoding_generated_prons_pruned.txt $dir/g2p_pron_candidates_pruned.txt \
```

Comment:
for lines like this, there always needs to be a comment explaining what is going on.
also, the name 'dict_combined2' is not very informative.. I'm not sure what to suggest though.